src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java

```diff
@@ -213,73 +262,99 @@ protected boolean requiresReference() {
-                final ReferenceSequenceFileWalker referenceSequenceFileWalker = new ReferenceSequenceFileWalker(REFERENCE_SEQUENCE);
-                final PeekableIterator<VariantContext> vcfIterator = new PeekableIterator<>(
-                        VCF == null ? Collections.emptyIterator() : new VCFFileReader(VCF, true).iterator())
+            final SamReader sam = SamReaderFactory.makeDefault()
+                    .referenceSequence(REFERENCE_SEQUENCE)
+                    .setOption(SamReaderFactory.Option.INCLUDE_SOURCE_IN_RECORDS, true)
+                    .open(INPUT);
+
+            final ReferenceSequenceFileWalker referenceSequenceFileWalker = new ReferenceSequenceFileWalker(REFERENCE_SEQUENCE);
+
+            final PeekableIterator<VariantContext> vcfIterator = new PeekableIterator<>(
+                    VCF == null ? Collections.emptyIterator() : new VCFFileReader(VCF, true).iterator())
         ) {
+            // Go through our reads and variants now:
+            final SamLocusAndReferenceIterator iterator = createSamLocusAndReferenceIterator(sam, referenceSequenceFileWalker);
 
-            final SAMSequenceDictionary sequenceDictionary = referenceSequenceFileWalker.getSequenceDictionary();
-            if (sam.getFileHeader().getSortOrder() != SAMFileHeader.SortOrder.coordinate) {
-                throw new PicardException("Input BAM must be sorted by coordinate");
-            }
+            log.info("Really starting iteration now.");
+            for (final SAMLocusAndReference info : iterator) {
+                if (random.nextDouble() > PROBABILITY) {
+                    continue;
+                }
+                nTotalLoci++;
+
+                // while there is a next (non-filtered) variant and it is before the locus, advance the pointer.
+                if (advanceIteratorAndCheckLocus(vcfIterator, info.getLocus(), sequenceDictionary)) {
+                    log.debug(String.format("Skipping locus from VCF: %s", vcfIterator.peek().toStringWithoutGenotypes()));
+                    nSkippedLoci++;
+                    continue;
+                }
 
-            sequenceDictionary.assertSameDictionary(sam.getFileHeader().getSequenceDictionary());
+                addLocusBases(aggregatorList, info);
 
-            final IntervalList regionOfInterest = getIntervals(sequenceDictionary);
+                nProcessedLoci++;
+                progressLogger.record(info.getLocus().getSequenceName(), info.getLocus().getPosition());
 
-            log.info("Getting SamLocusIterator");
+                if (MAX_LOCI != 0 && nProcessedLoci >= MAX_LOCI) {
+                    log.warn("Early stopping due to having processed MAX_LOCI loci.");
+                    break;
+                }
+            }
 
-            final SamLocusIterator samLocusIterator = new SamLocusIterator(sam, regionOfInterest);
+        } catch (IOException e) {
+            log.error(e, "A problem occurred:");
```

Comment:
I know you just copied the code..could you fix this by including the error message from the exception? here and below as well.

Reply:
Fixed.